File: ariprog.cpp
/*
ID: totheso1
LANG: C++14
TASK: ariprog
*/

#include <algorithm>
#include <fstream>
#include <iterator>
#include <utility>
#include <vector>

using namespace std;

static ifstream fin("ariprog.in");
static ofstream fout("ariprog.out");

#define endl '\n'

template <typename T>
T fin_get() {
  return *istream_iterator<T>(fin);
}

template <typename C>
C fin_get_collection(const int size) {
  C ret;
  copy_n(istream_iterator<typename C::value_type>(fin), size,
         back_inserter(ret));
  return ret;
}

int main() {
  const int LENGTH_OF_PROGRESSION = fin_get<int>();
  const int MAX_FACTOR = fin_get<int>();
  const int MAX_NUMBER = MAX_FACTOR * MAX_FACTOR * 2;

  vector<bool> available(MAX_NUMBER + 1, false);
  vector<int> number_collection;
  number_collection.reserve(MAX_NUMBER + 1);
  for (int i = 0; i <= MAX_FACTOR; ++i) {
    for (int j = 0; j <= i; ++j) {
      const int number = i * i + j * j;
      if (!available[number]) {
        available[number] = true;
        number_collection.push_back(number);
      }
    }
  }
  sort(begin(number_collection), end(number_collection));

  const int back = number_collection.back();
  vector<pair<int, int>> result;
  for (auto i = cbegin(number_collection), e = cend(number_collection); i < e;
       ++i) {
    const int num_a = *i;
    for (auto j = i + 1; j < e; ++j) {
      const int num_b = *j;
      const int diff = num_b - num_a;
      const int last_num = num_a + diff * (LENGTH_OF_PROGRESSION - 1);
      if (back < last_num) {
        break;
      }
      bool found = true;
      for (int next_num = num_b + diff; next_num <= last_num;
           next_num += diff) {
        if (!available[next_num]) {
          found = false;
          break;
        }
      }
      if (found) {
        result.emplace_back(diff, num_a);
      }
    }
  }
  sort(begin(result), end(result));

  if (result.empty()) {
    fout << "NONE" << endl;
  } else {
    for (const auto &p : result) {
      fout << p.second << ' ' << p.first << endl;
    }
  }
}

File: ariprog.py
"""
ID: totheso1
LANG: PYTHON3
TASK: ariprog
NOTE:
    This task can't be done in python in time limit. Use C++14 instead.
    The following program is correct,
    but will take about 10 seconds and get TLE when M is 250.
"""

from time import time

start_time = time()

from itertools import islice
from typing import Any, List

fin = open("ariprog.in", "r")
fout = open("ariprog.out", "w")


def fprint(*args: Any, **kwargs: Any) -> None:
    print(*args, file=fout, **kwargs)


# Code start
LENGTH_OF_PROGRESSION = int(fin.readline())
MAX_FACTOR = int(fin.readline())
MAX_NUMBER = MAX_FACTOR * MAX_FACTOR * 2

available = [False] * (MAX_NUMBER + 1)
number_collection: List[int] = []
for i in range(MAX_FACTOR + 1):
    for j in range(i + 1):
        number = i * i + j * j
        if not available[number]:
            available[number] = True
            number_collection.append(number)
number_collection.sort()

back = number_collection[-1]
result = []
for i, num_a in enumerate(number_collection):
    for j, num_b in enumerate(islice(number_collection, i + 1, None)):
        diff = num_b - num_a
        last_num = num_a + diff * (LENGTH_OF_PROGRESSION - 1)
        if back < last_num:
            break
        for next_num in range(num_b + diff, last_num + 1, diff):
            if not available[next_num]:
                break
        else:
            result.append((diff, num_a))
result.sort()

if not result:
    fprint("NONE")
else:
    for diff, num_a in result:
        fout.write(f"{num_a} {diff}\n")
# Code end
print(time() - start_time)
